File: Java/Sorting Algorithms/Merge Sort/MergeSort.java
class MergeSort{

	int array[]={10,20,51,11,42,32,78,95,15};


	public void merge(int left, int mid, int right){

		int temp[]=new int[right-left+1];
		int i=left;
		int j=mid+1;
		int k=0;

		while(i<=mid && j<=right)
		{
			if(array[i] <= array[j])
			{
				temp[k]=array[i];
				k++;
				i++;
			}

			else
			{
				temp[k]=array[j];
				k++;
				j++;
			}
		}

		while(j<=right)
		{
			temp[k++]=array[j++];
		}

		while(i<=mid)
		{
			temp[k++]=array[i++];
		}

		for(k=0; k<temp.length; k++)
		{
			array[left+k]=temp[k];
		}

	}

	public void mergeSort(int left, int right)
	{
		if(left<right){

			int mid =(left + right)/2;

			mergeSort(left,mid);
			mergeSort(mid+1,right);
			merge(left,mid,right);
		}
	}

	public void print()
	{
		for(int i=0; i<array.length; i++)
		{
			System.out.print(array[i]+" ");
		}
		System.out.println();

	}

	public static void main(String args[])
	{
		MergeSort ob =new MergeSort();

		ob.print();
		//ob.merge(0,5,8);
		ob.mergeSort(0,8);

		ob.print();


	}
}

File: JavaScript/MergeSort/Mergesort.js
function merge(left_part,right_part) 
{
	var i = 0;
	var j = 0;
	var results = [];

	while (i < left_part.length || j < right_part.length) {
		if (i === left_part.length) {
			// j is the only index left_part
			results.push(right_part[j]);
			j++;
		} 
      else if (j === right_part.length || left_part[i] <= right_part[j]) {
			results.push(left_part[i]);
			i++;
		} else {
			results.push(right_part[j]);
			j++;
		}
	}
	return results;
}

function mergeSort(arr) {
    if(arr.length <= 1) return arr;
    var mid = Math.floor(arr.length/2);
    var left = mergeSort(arr.slice(0,mid));
    var right = mergeSort(arr.slice(mid));
    return merge(left, right);
}


console.log(mergeSort([2, 3, 3, 4, 6, 7, 12]));
